diff --git a/CMakeLists.txt b/CMakeLists.txt
index 982535b7..2a9e2d08 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -116,17 +116,14 @@ if(USE_SWIG)
 endif()
 
 set(EIGEN_DIR "${PROJECT_SOURCE_DIR}/external_libs/eigen")
-include_directories(${EIGEN_DIR})
 
 # See https://gitlab.com/libeigen/eigen/-/blob/master/COPYING.README
 add_definitions(-DEIGEN_MPL2_ONLY)
 add_definitions(-DEIGEN_DONT_PARALLELIZE)
 
 set(FAST_DOUBLE_PARSER_INCLUDE_DIR "${PROJECT_SOURCE_DIR}/external_libs/fast_double_parser/include")
-include_directories(${FAST_DOUBLE_PARSER_INCLUDE_DIR})
 
 set(FMT_INCLUDE_DIR "${PROJECT_SOURCE_DIR}/external_libs/fmt/include")
-include_directories(${FMT_INCLUDE_DIR})
 
 if(__BUILD_FOR_R)
     find_package(LibR REQUIRED)
@@ -785,13 +782,4 @@ if(INSTALL_HEADERS)
       DIRECTORY ${LightGBM_HEADER_DIR}/LightGBM
       DESTINATION ${CMAKE_INSTALL_PREFIX}/include
     )
-    install(
-      FILES ${FAST_DOUBLE_PARSER_INCLUDE_DIR}/fast_double_parser.h
-      DESTINATION ${CMAKE_INSTALL_PREFIX}/include/LightGBM/utils
-    )
-    install(
-      DIRECTORY ${FMT_INCLUDE_DIR}/
-      DESTINATION ${CMAKE_INSTALL_PREFIX}/include/LightGBM/utils
-      FILES_MATCHING PATTERN "*.h"
-    )
 endif()
