# XOOS Secondary Analysis

## Introduction

XOOS (SB**X** **O**ptimized **O**pen **S**ource) analysis tools are a comprehensive suite of bioinformatics tools designed and optimized for processing SBX data generated on the AXELIOS platform.
This **early access release** focuses specifically on germline small variant calling (SNVs and indels) for SBX-Duplex (SBX-D) WGS data, transforming aligned reads into high-quality, analysis-ready variant calls.

As an **open source solution**, XOOS is freely available for the scientific community to adopt, customize, and enhance.
We encourage users to:

- **Adapt** the tools to your specific needs
- **Configure** parameters and workflows for your computational environment
- **Extend** functionality with additional tools and analysis modules
- **Contribute back** improvements, bug fixes, and new features to benefit the SBX community

All XOOS analysis tools source code is hosted on [GitHub](https://github.com/Roche-AXELIOS/XOOS), please see [Releases](https://github.com/Roche-AXELIOS/XOOS/releases) for latest release notes, and [Packages](https://github.com/orgs/Roche-AXELIOS/packages?repo_name=XOOS) for pre-built Docker images.

## Disclaimer

Unless required by applicable law or agreed to in writing, XOOS analysis tools are provided on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied, including, without limitation, any warranties or conditions of TITLE, NON-INFRINGEMENT, MERCHANTABILITY, or FITNESS FOR A PARTICULAR PURPOSE.
You are solely responsible for determining the appropriateness of using or redistributing the Work and assume any risks associated with your use of XOOS analysis tools.
